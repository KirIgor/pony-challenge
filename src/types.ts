import { Map as iMap } from 'immutable';

export enum Role {
	PONY = 'pony',
	DOMOKUN = 'domokun',
	EXIT = 'exit',
	NONE = 'none'
}

export enum Side {
	LEFT = 1,
	RIGHT = 2,
	TOP = 4,
	BOTTOM = 8
}

export interface Point {
	x: number;
	y: number;
}

export enum PonyName {
	TWILIGHT_SPARKLE = 'Twilight Sparkle',
	APPLEJACK = 'Applejack',
	FLUTTERSHY = 'Fluttershy',
	RARITY = 'Rarity',
	PINKIE_PIE = 'Pinkie Pie',
	RAINBOW_DASH = 'Rainbow Dash',
	SPIKE = 'Spike'
}

export enum Direction {
	WEST = 'west',
	EAST = 'east',
	NORTH = 'north',
	SOUTH = 'south'
}

export type CharactersPosition = iMap<Role, Point>;

// sides is bitwise OR of sides
export interface Blueprint {
	[index: number]: {
		[index: number]: {
			sides: number;
		};
	};
}

export interface MazeProps {
	width: number;
	height: number;
	charactersPosition: CharactersPosition;
	blueprint: Blueprint;
}

export type Position = [number];
export interface GameState {
	pony: Position;
	domokun: Position;
	'end-point': Position;
	size: [number, number];
	difficulty: number;
	data: string[][];
	maze_id: string;
	'game-state': {
		state: string;
		'state-result': string;
	};
}
