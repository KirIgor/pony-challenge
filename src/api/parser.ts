import {
	MazeProps,
	GameState,
	Blueprint,
	CharactersPosition,
	Direction,
	Side,
	Position,
	Role,
	Point
} from '../types';

import { range } from '../utils/helper';
import { Map as iMap } from 'immutable';

const directionToSide = (direction: Direction): Side => {
	switch (direction) {
		case Direction.WEST: {
			return Side.LEFT;
		}
		case Direction.EAST: {
			return Side.RIGHT;
		}
		case Direction.NORTH: {
			return Side.TOP;
		}
		case Direction.SOUTH: {
			return Side.BOTTOM;
		}
	}
};

const parseBlueprint = (data: string[][], [width, height]: [number, number]): Blueprint =>
	range(height).map(i =>
		data.slice(i * width, i * width + width).map(directions => ({
			sides: directions
				.map(direction => directionToSide(Direction[direction.toUpperCase()]))
				.reduce((acc, cur) => (acc |= cur), 0)
		}))
	);

const parsePosition = (pos: Position, width: number): Point => ({
	x: pos[0] % width,
	y: Math.floor(pos[0] / width)
});

const parseCharactersPosition = (
	[width, height]: [number, number],
	pony: Position,
	domokun: Position,
	exit: Position
): CharactersPosition =>
	iMap<Role, Point>().withMutations(map =>
		map
			.set(Role.PONY, parsePosition(pony, width))
			.set(Role.DOMOKUN, parsePosition(domokun, width))
			.set(Role.EXIT, parsePosition(exit, width))
	);

export const parseGameState = (state: GameState): MazeProps => ({
	width: state.size[0],
	height: state.size[1],
	charactersPosition: parseCharactersPosition(
		state.size,
		state.pony,
		state.domokun,
		state['end-point']
	),
	blueprint: parseBlueprint(state.data, state.size)
});
