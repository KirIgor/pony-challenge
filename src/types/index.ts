import { RecordFactory, StaticallyTypedRecord } from '../utils/immutable_helper';
import { Map as iMap, List as iList } from 'immutable';

export enum Role {
	PONY = 'pony',
	DOMOKUN = 'domokun',
	EXIT = 'exit',
	NONE = 'none'
}

export enum Side {
	LEFT = 1,
	RIGHT = 2,
	TOP = 4,
	BOTTOM = 8
}

interface IPoint {
	x: number;
	y: number;
}

export const point = RecordFactory<IPoint>({
	x: 0,
	y: 0
});

export class Point extends point implements IPoint, StaticallyTypedRecord<IPoint> {
	x: number;
	y: number;
	constructor(props: IPoint) {
		super(props);
	}
}

export enum PonyName {
	TWILIGHT_SPARKLE = 'Twilight Sparkle',
	APPLEJACK = 'Applejack',
	FLUTTERSHY = 'Fluttershy',
	RARITY = 'Rarity',
	PINKIE_PIE = 'Pinkie Pie',
	RAINBOW_DASH = 'Rainbow Dash'
}

export enum Direction {
	WEST = 'west',
	EAST = 'east',
	NORTH = 'north',
	SOUTH = 'south'
}

export type CharactersPosition = iMap<Role, Point>;

// sides is bitwise OR of sides
interface IBlueprintRecord {
	sides: number;
}

const blueprintRecord = RecordFactory<IBlueprintRecord>({
	sides: 0
});

export class BlueprintRecord extends blueprintRecord
	implements IBlueprintRecord, StaticallyTypedRecord<IBlueprintRecord> {
	sides: number;
	constructor(props: IBlueprintRecord) {
		super(props);
	}
}

export type Blueprint = iList<iList<BlueprintRecord>>;

interface IGameState {
	mazeId: string;
	width: number;
	height: number;
	charactersPosition: CharactersPosition;
	blueprint: Blueprint;
}

const gameState = RecordFactory<IGameState>({
	mazeId: '',
	width: 0,
	height: 0,
	charactersPosition: iMap<Role, Point>(),
	blueprint: iList<iList<BlueprintRecord>>()
});

export class GameState extends gameState implements IGameState, StaticallyTypedRecord<IGameState> {
	mazeId: string;
	width: number;
	height: number;
	charactersPosition: CharactersPosition;
	blueprint: Blueprint;
	constructor(props: IGameState) {
		super(props);
	}
}

export type Position = [number];
export interface APIState {
	pony: Position;
	domokun: Position;
	'end-point': Position;
	size: [number, number];
	difficulty: number;
	data: string[][];
	maze_id: string;
	'game-state': {
		state: string;
		'state-result': string;
	};
}

export enum RainbowType {
	NORTH_TO_EAST = 'NORTH_TO_EAST',
	EAST_TO_SOUTH = 'EAST_TO_SOUTH',
	SOUTH_TO_WEST = 'SOUTH_TO_WEST',
	WEST_TO_NORTH = 'WEST_TO_NORTH',
	NORTH_RED_TO_LEFT = 'NORTH_RED_TO_LEFT',
	NORTH_RED_TO_RIHGT = 'NORTH_RED_TO_RIGHT',
	WEST_RED_TO_TOP = 'WEST_RED_TO_TOP',
	WEST_RED_TO_BOTTOM = 'WEST_RED_TO_BOTTOM',
	NONE = 'NONE'
}
export type RainbowPosition = { x: number; y: number; direction: Direction };
export type RainbowPath = [] | [RainbowPosition] | [RainbowPosition, RainbowPosition];

export enum BorderConnection {
	TOP_LEFT = 1,
	TOP_RIGHT = 2,
	BOTTOM_LEFT = 4,
	BOTTOM_RIGHT = 8
}
