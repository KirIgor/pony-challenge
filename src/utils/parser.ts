import { fromJS } from 'immutable';

import {
	GameState,
	APIState,
	Blueprint,
	CharactersPosition,
	Direction,
	Side,
	Position,
	Role,
	Point,
	GameStatus
} from '../types/index';
import { range } from './helper';
import { Map as iMap } from 'immutable';

const directionToSide = (direction: Direction): Side => {
	switch (direction) {
		case Direction.WEST: {
			return Side.LEFT;
		}
		case Direction.EAST: {
			return Side.RIGHT;
		}
		case Direction.NORTH: {
			return Side.TOP;
		}
		case Direction.SOUTH: {
			return Side.BOTTOM;
		}
	}
};

export const parseBlueprint = ({ data, size: [width, height] }: APIState): Blueprint =>
	fromJS(
		range(height).map(i =>
			data.slice(i * width, i * width + width).map((directions, j) => ({
				sides: directions
					.map(direction => directionToSide(Direction[direction.toUpperCase()]))
					.reduce(
						(acc, cur) => (acc |= cur),
						(+(i === height - 1) * Side.BOTTOM) | (+(j === width - 1) * Side.RIGHT)
					)
			}))
		)
	);

const parsePosition = (pos: Position, width: number): Point =>
	new Point({
		x: pos[0] % width,
		y: Math.floor(pos[0] / width)
	});

export const parseCharactersPosition = ({
	size: [width, _],
	pony,
	domokun,
	['end-point']: exit
}: APIState): CharactersPosition =>
	iMap<Role, Point>().withMutations(map =>
		map
			.set(Role.PONY, parsePosition(pony as [number], width))
			.set(Role.DOMOKUN, parsePosition(domokun as [number], width))
			.set(Role.EXIT, parsePosition(exit as [number], width))
	);

export const parseGameStatus = (state: APIState): GameStatus => {
	switch (state['game-state'].state) {
		case 'won':
			return GameStatus.WIN;
		case 'over':
			return GameStatus.LOSE;
		default:
			return GameStatus.ACTIVE;
	}
};

export const parseGameState = (state: APIState): GameState =>
	new GameState({
		mazeId: state.maze_id,
		width: state.size[0],
		height: state.size[1],
		gameStatus: parseGameStatus(state),
		charactersPosition: parseCharactersPosition(state),
		blueprint: parseBlueprint(state)
	});
