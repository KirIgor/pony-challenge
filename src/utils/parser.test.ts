import { APIState } from '../types';
import { parseGameState } from './parser';

const stateMock: APIState = {
	pony: [3],
	domokun: [7],
	'end-point': [8],
	size: [3, 3],
	difficulty: 0,
	data: [
		['west', 'north'],
		['west', 'north'],
		['north'],
		['north'],
		['north'],
		['north'],
		['north'],
		['north'],
		['west', 'north']
	],
	maze_id: 'c6d8aefa-23fb-42a3-8eeb-347ab807de0d',
	'game-state': {
		state: 'Active',
		'state-result': 'Successfully created'
	}
};

describe('test game state parser', () => {
	it('should match snapshot', () => {
		expect(parseGameState(stateMock)).toMatchSnapshot();
	});
});
