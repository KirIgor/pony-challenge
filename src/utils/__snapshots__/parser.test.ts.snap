// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`game state parser should match snapshot 1`] = `
Object {
  "blueprint": Immutable.List [
    Immutable.List [
        Immutable.Map {
            sides: 5,
        },
        Immutable.Map {
            sides: 5,
        },
        Immutable.Map {
            sides: 6,
        },
    ],
    Immutable.List [
        Immutable.Map {
            sides: 4,
        },
        Immutable.Map {
            sides: 4,
        },
        Immutable.Map {
            sides: 6,
        },
    ],
    Immutable.List [
        Immutable.Map {
            sides: 12,
        },
        Immutable.Map {
            sides: 12,
        },
        Immutable.Map {
            sides: 15,
        },
    ],
],
  "charactersPosition": Immutable.Map {
    pony: Object {
        "x": 0,
        "y": 1,
      },
    domokun: Object {
        "x": 1,
        "y": 2,
      },
    exit: Object {
        "x": 2,
        "y": 2,
      },
},
  "gameStatus": "ACTIVE",
  "height": 3,
  "mazeId": "c6d8aefa-23fb-42a3-8eeb-347ab807de0d",
  "width": 3,
}
`;
